env/env_encryption.cc

```diff
@@ -361,9 +413,152 @@ Status EncryptedRandomRWFile::Close() { return file_->Close(); }
 
 // EncryptedEnv implements an Env wrapper that adds encryption to files stored
 // on disk.
-class EncryptedEnv : public EnvWrapper {
+class EncryptedEnvImpl : public EnvWrapper {
+  Status GetWritableProvider(const std::string& /*fname*/,
+                             EncryptionProvider** result) {
+    if (provider_) {
+      *result = provider_.get();
+      return Status::OK();
+    } else {
+      *result = nullptr;
+      return Status::NotFound("No WriteProvider specified");
+    }
+  }
+
+  Status GetReadableProvider(const std::string& /*fname*/,
+                             EncryptionProvider** result) {
+    if (provider_) {
+      *result = provider_.get();
+      return Status::OK();
+    } else {
+      *result = nullptr;
+      return Status::NotFound("No Provider specified");
+    }
+  }
+
+  template <class TypeFile>
+  Status CreateWritableCipherStream(
+      const std::string& fname, const std::unique_ptr<TypeFile>& underlying,
+      const EnvOptions& options, size_t* prefix_length,
+      std::unique_ptr<BlockAccessCipherStream>* stream) {
+    EncryptionProvider* provider = nullptr;
```

Comment:
assume this comment everywhere ... should it be std::shared_ptr<EncryptionProvider> ?